network/response_parser.c

```diff
@@ -20,80 +22,100 @@ int parse_uri(char *response, http_response *result);
  * @param index
  * @return index of body start in response
  */
-int parse_header(char *response, http_response *result, int index);
+int parse_header(char *line, http_response *result, int index);
 
 /**
  * parse http response to
  * @param response
  * @return http response
  */
 http_response *parse(char *response) {
-	http_response *result = malloc(sizeof(http_response) + 4 * sizeof(char *));
-	int header_index = parse_uri(response, result);
-	int body_index = parse_header(response, result, header_index);
-	result->body = string_from_to(response, body_index, length_pointer_char(response) - 1);
-	return result;
+    char** response_split = string_split(response, "\r\n");
+    int response_split_length = length_pointer_pointer_char(response_split);
+    http_response *result = malloc(sizeof(http_response) + 4 * sizeof(char *));
+    result->header_quantity = count_header(response);
+    if (response_split_length < 2) {
+        free_pointer_pointer_char(response_split);
+        return result;
+    }
+
+    if (result->header_quantity == 0 || !parse_uri(response_split[0], result)) {
+        return result;
+    }
+    register int index;
+    for (index = 0; index < result->header_quantity; index ++) {
+        if (!parse_header(response_split[index + 1], result, index)) {
+            return result;
+        }
+    }
+
+    result->body = strdup("");
+    if (result->header_quantity + 1 < response_split_length) {
+        char* pointer_holder;
+        free(result->body);
+        result->body = strdup(response_split[result->header_quantity + 1]);
+        for (index = result->header_quantity + 2; index < response_split_length; index ++) {
+            pointer_holder = result->body;
+            asprintf(&result->body, "%s\r\n%s", result->body, response_split[index]); // append all data remain to body
+            free(pointer_holder);
+        }
+    }
```

Comment:
Function will return `result`. That's reason we could not free anything in `result`